### PR TITLE
storage: actually ensure the topic exists when purifying kafka sources

### PR DESCRIPTION
**Pull request by guswynn**
This must have been lost in a rebase somewhere. Im genuinely astounded by this, as if you dont specify start offsets, you get no validation on your kafka source working...

Note this would have made the race in incident-92 wayyy less likely

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by petrosagg — APPROVED** [review]

**Review by benesch — APPROVED** [review]
Ooof! Good catch.

File: src/sql/src/kafka_util.rs [hunk 1]
```diff
@@ -224,3 +224,28 @@ where
...
+{
+    task::spawn_blocking(|| format!("kafka_ensure_topic_exists:{topic}"), {
+        let topic = topic.to_string();
+        move || {
+            mz_kafka_util::client::get_partitions(
+                consumer.as_ref().client(),
+                &topic,
+                fetch_metadata_timeout,
+            )
+            .map_err(|e| sql_err!("{}", e))?;
```
Comment (ParkMyCar) [inline src/sql/src/kafka_util.rs:245]: I'm not sure we want to map this directly to a SQL error, do you mind adding a new "Topic Missing" variant to `PlanError`? Reason being, it looks like `get_paritions` can return errors from Kafka itself which maybe could contain sensitive information? 

If we do this conversion using `sql_err!` elsewhere then feel free to ignore me, but this seems like a small thing that could bite us in the future

**Review by ParkMyCar — APPROVED** [review]

File: src/sql/src/kafka_util.rs (hunk 1 above)
Comment (guswynn) [reply src/sql/src/kafka_util.rs:245]: Yeah we do this for all the kafka stuff; we do parse the kafka errors to make better errors, but then just convert those to generic `sql_err`'s. We could make kafka error type variants in `PlanError` if we want, but that would have to be followup

**Comment by guswynn** [comment]
Oh rip, tests that use sources on top of sink progress topics need to pre-create the progress topic now

File: ci/test/cargo-test/mzcompose.py [hunk 2]
```diff
@@ -28,6 +28,7 @@
             "KAFKA_ADVERTISED_LISTENERS=HOST://localhost:30123,PLAINTEXT://kafka:9092",
             "KAFKA_LISTENER_SECURITY_PROTOCOL_MAP=HOST:PLAINTEXT,PLAINTEXT:PLAINTEXT",
         ],
+        auto_create_topics=True
```
Comment (petrosagg) [inline ci/test/cargo-test/mzcompose.py:31]: This doesn't look right, we should fix up the specific tests that need a topic by creating it

**Review by ParkMyCar — APPROVED** [review]

**Pull request merged**